pkg/stargate/affinity.go

```diff
@@ -0,0 +1,107 @@
+package stargate
+
+import (
+	cassdcapi "github.com/k8ssandra/cass-operator/operator/pkg/apis/cassandra/v1beta1"
+	"github.com/k8ssandra/k8ssandra-operator/pkg/utils"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sort"
+)
+
+// Deprecated label name used in cass-operator.
+// See pkg/reconciliation/construct_statefulset.go
+const zoneLabel = "failure-domain.beta.kubernetes.io/zone"
+
+// computeNodeAffinityLabels computes a set of labels to select worker nodes for the Stargate pods.
+// We use the same labels that were used to select worker nodes for the data pods.
+// See cass-operator, pkg/reconciliation/construct_podtemplatespec.go
+func computeNodeAffinityLabels(dc *cassdcapi.CassandraDatacenter, rackName string) map[string]string {
+	var labels map[string]string
+	racks := dc.GetRacks()
+	for _, rack := range racks {
+		if rack.Name == rackName {
+			labels = utils.MergeMap(rack.NodeAffinityLabels, dc.Spec.NodeAffinityLabels)
```

Comment:
@jsanda my comment got deleted by a force-push so I'm reposting so that we can resume the conversation:

> Wanted to mention that this comes from cassdc api. As is, DC spec overrides rack spec. This seems in contradiction with K8ssandra's hierarchical principles but I didn't want to change the logic.

From @jsanda :

> That's a good point. Since it is the cluster controller who is creating the CassandraDatacenters, we can avoid using the .Spec.NodeAffinityLabels property. There is one caveat though. At some point we will need to adding an existing CassandraDatacenter to a K8ssandraCluster, particularly for migration scenarios. Maybe we should raise an issue in cass-operator for changing the behavior.

I agree to raise an issue, I will take care of it.

But now that you mention this, I wonder if we shouldn't include the following properties in `CassandraDatacenterTemplateSpec`:

* `NodeAffinityLabels map[string]string `
* `NodeSelector map[string]string`
* `Tolerations []corev1.Toleration`
* `AllowMultipleNodesPerWorker bool`

My understanding is that the above properties greatly influence pod scheduling, and I guess that most users would eventually like to see them in the K8ssandraCluster API.

Note: there is also the field `PodTemplateSpec *corev1.PodTemplateSpec` in `CassandraDatacenter` – but atm I think we can avoid exposing it.

Reply:
Here is the issue: https://github.com/k8ssandra/cass-operator/issues/162.

Reply:
Those properties are definitely going to be added in `CassandraDatacenterTemplateSpec`. I expect some or all of that to be done as part of #87. It

Thanks for creating the ticket!